Topo: Add version support to GetTopologyPath

mattlord opened the pull request:
## Description

This PR adds the ability to view older/previous versions of topology keys. This can be used to follow changes over time and to revert to previous values.

ℹ️ Notes:
  - This is only supported in the `etcd2topo` implementation. This is because [etcd is the only topo server implementation that stores multiple versions/revisions of a key](https://github.com/vitessio/vitess/pull/15933#issuecomment-2118641068) and allows you to request a specific one. For ZooKeeper and Consul we display the value — which is a modification counter — and return an intuitive error if you use the flag:
```
❯ vtctldclient GetTopologyPath /global/routing_rules/keyspace/Rules --data-as-json --version=1
E0518 01:32:36.674734   45693 main.go:56] rpc error: code = Unknown desc = no such topology implementation GetVersion not supported in ZK2 topo

❯ vtctldclient GetTopologyPath /global/routing_rules/keyspace/Rules --data-as-json --version=1
E0518 01:33:55.596021   46502 main.go:56] rpc error: code = Unknown desc = no such topology implementation GetVersion not supported in consul topo
```
  - There is no way to "see all available versions of a node", so you can only step back in time and look at previous values for the node at previous keyspace revisions (see examples below)

Example usage:
```
pushd examples/local

./101_initial_cluster.sh

❯ vtctldclient ApplyKeyspaceRoutingRules --rules '{"rules":[{"from_keyspace":"commerce", "to_keyspace":"customer"}]}'
{
  "keyspace_routing_rules": {
    "rules": [
      {
        "from_keyspace": "commerce",
        "to_keyspace": "customer"
      }
    ]
  }
}

❯ vtctldclient GetTopologyPath /global/routing_rules/keyspace/Rules
{
  "name": "Rules",
  "path": "/global/routing_rules/keyspace/Rules",
  "data": "rules:{from_keyspace:\"commerce\" to_keyspace:\"customer\"}",
  "children": [],
  "version": "36"
}

❯ vtctldclient ApplyKeyspaceRoutingRules --rules '{}'
{
  "keyspace_routing_rules": {
    "rules": []
  }
}

❯ vtctldclient GetTopologyPath /global/routing_rules/keyspace/Rules --data-as-json
{}

❯ vtctldclient GetTopologyPath /global/routing_rules/keyspace/Rules
{
  "name": "Rules",
  "path": "/global/routing_rules/keyspace/Rules",
  "data": "",
  "children": [],
  "version": "39"
}

❯ vtctldclient GetTopologyPath /global/routing_rules/keyspace/Rules --data-as-json --version=38
{
  "rules": [
    {
      "from_keyspace": "commerce",
      "to_keyspace": "customer"
    }
  ]
}

❯ vtctldclient ApplyKeyspaceRoutingRules --rules "$(vtctldclient GetTopologyPath /global/routing_rules/keyspace/Rules --data-as-json --version=38)"
{
  "keyspace_routing_rules": {
    "rules": [
      {
        "from_keyspace": "commerce",
        "to_keyspace": "customer"
      }
    ]
  }
}

❯ vtctldclient GetTopologyPath /global/routing_rules/keyspace/Rules
{
  "name": "Rules",
  "path": "/global/routing_rules/keyspace/Rules",
  "data": "rules:{from_keyspace:\"commerce\" to_keyspace:\"customer\"}",
  "children": [],
  "version": "43"
}

❯ vtctldclient UpdateThrottlerConfig --enable --custom-query="select 1 from foo" --throttle-app="vplayer" commerce

❯ vtctldclient GetTopologyPath /global/keyspaces/commerce/Keyspace --data-as-json
{
  "durability_policy": "semi_sync",
  "sidecar_db_name": "_vt",
  "throttler_config": {
    "custom_query": "select 1 from foo",
    "enabled": true,
    "throttled_apps": {
      "vplayer": {
        "expires_at": {
          "nanoseconds": 774788000,
          "seconds": "1715964511"
        },
        "name": "vplayer",
        "ratio": 1
      }
    }
  }
}
```

## Related Issue(s)

  - Fixes: https://github.com/vitessio/vitess/issues/15968

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation PR: https://github.com/vitessio/website/pull/1757

mattlord commented on go/vt/topo/decode.go:
```diff
@@ -53,6 +53,11 @@ func DecodeContent(filename string, data []byte, json bool) (string, error) {
 		p = new(topodatapb.SrvKeyspace)
 	case RoutingRulesFile:
 		p = new(vschemapb.RoutingRules)
+	case CommonRoutingRulesFile:
+		switch path.Base(dir) {
+		case "keyspace":
+			p = new(vschemapb.KeyspaceRoutingRules)
+		}
```
This was needed as a follow-up to: https://github.com/vitessio/vitess/pull/15807

This PR is also what allows you to easily revert to the previous value as we had toyed with the idea of returning the *old and new* values from the `ApplyKeyspaceRoutingRules` command for this purpose.

deepthi commented:
I don't like the idea of doing something that only works with etcd.
Also, how do you know what the available versions are?
What happens if someone is running with zookeeper (which we still support), and tries to use this feature?

mattlord commented:
> I don't like the idea of doing something that only works with etcd. 

ZooKeeper does not support this feature generally AFAIK. It's also worth noting that etcd2 is the only implementation that is officially / fully supported AFAIK. IMO it's a nice feature to support where it's possible.

> Also, how do you know what the available versions are?

The etcd keyspace has a new revision each time the keyspace is changed. So you would see what the current version is for the key and you can step back from there. That's what I did in the examples in the PR description. So e.g. if you just made a change, you would use current version - 1 so see what value of the node/key was at the previous keyspace revision.

> What happens if someone is running with zookeeper (which we still support), and tries to use this feature?

You'll get a an err: `GetVersion not supported in ZK2 topo`: https://github.com/vitessio/vitess/pull/15933/files#diff-05359f832e44437ddcaef9416e6f0c0fd0997dc84104b166c0948c7570a2a15a

deepthi commented:
Ok, this seems reasonable given the answers to my questions. Will review.

mattlord commented:
@deepthi actually... it seems like ZooKeeper does support versions — there are virtually no docs so I didn't find anything, but... (after fixing the local example in: https://github.com/vitessio/vitess/pull/15933/commits/e5bef58d40044a80cea0f51808f1a26f1960b07e):
```
export TOPO=zk2
❯ ./101_initial_cluster.sh
...

❯ vtctldclient GetTopologyPath /global/keyspaces/commerce/Keyspace
{
  "name": "Keyspace",
  "path": "/global/keyspaces/commerce/Keyspace",
  "data": "durability_policy:\"semi_sync\" sidecar_db_name:\"_vt\"",
  "children": [],
  "version": "0"
}

❯ vtctldclient UpdateThrottlerConfig --enable --custom-query="select 1 from foo" --throttle-app="vplayer" commerce

❯ vtctldclient GetTopologyPath /global/keyspaces/commerce/Keyspace
{
  "name": "Keyspace",
  "path": "/global/keyspaces/commerce/Keyspace",
  "data": "durability_policy:\"semi_sync\" throttler_config:{enabled:true custom_query:\"select 1 from foo\" throttled_apps:{key:\"vplayer\" value:{name:\"vplayer\" ratio:1 expires_at:{seconds:1715971901 nanoseconds:349687000}}}} sidecar_db_name:\"_vt\"",
  "children": [],
  "version": "1"
}

❯ vtctldclient UpdateThrottlerConfig --enable --custom-query="select 1 from foobar" --throttle-app="vplayer" commerce

❯ vtctldclient GetTopologyPath /global/keyspaces/commerce/Keyspace
{
  "name": "Keyspace",
  "path": "/global/keyspaces/commerce/Keyspace",
  "data": "durability_policy:\"semi_sync\" throttler_config:{enabled:true custom_query:\"select 1 from foobar\" throttled_apps:{key:\"vplayer\" value:{name:\"vplayer\" ratio:1 expires_at:{seconds:1715971909 nanoseconds:518421000}}}} sidecar_db_name:\"_vt\"",
  "children": [],
  "version": "2"
}
```

And we don't in fact get an error when trying to get the version:
```
❯ vtctldclient GetTopologyPath /global/keyspaces/commerce/Keyspace --version 1
{
  "name": "Keyspace",
  "path": "/global/keyspaces/commerce/Keyspace",
  "data": "",
  "children": [],
  "version": "0"
}
```

So that's not right... I'll put this back in draft then and try to add zk2 and consul support. And at the very least, return an error as expected if it's not supported.

mattlord commented:
Confirmed that ZooKeeper does not support getting a specific version of a node/key this as it does not actually hold revisions/versions of a key: https://github.com/z-division/go-zookeeper/blob/v1.0.0/zk/conn.go#L1129-L1166

Rather the version in ZK is merely a counter that is incremented on each Set: https://github.com/z-division/go-zookeeper/blob/v1.0.0/zk/structs.go#L36

So for ZK2 we can only provide a clear and meaningful error from the client. 

I'll look into consul next.

rohit-nayak-ps commented on go/vt/vtctl/grpcvtctldserver/server.go:
```diff
@@ -5062,8 +5065,8 @@ func StartServer(s *grpc.Server, env *vtenv.Environment, ts *topo.Server) {
 
-// getTopologyCell is a helper method that returns a topology cell given its path.
-func (s *VtctldServer) getTopologyCell(ctx context.Context, cellPath string) (*vtctldatapb.TopologyCell, error) {
+// getTopologyKey is a helper method that returns a topology key given its path.
```
Was this comment change intentional? Doesn't match the function name.

mattlord replied:
At the time it was, but I'll revert it as cell is used throughout and it also makes sense in this context.

rohit-nayak-ps commented on go/vt/vtctl/grpcvtctldserver/server.go:
```diff
@@ -5080,16 +5083,41 @@ func (s *VtctldServer) getTopologyCell(ctx context.Context, cellPath string) (*v
+	if version != 0 {
+		data, err := conn.GetVersion(ctx, relativePath, version)
+		if err == nil {
+			result, err := topo.DecodeContent(relativePath, data, asJSON)
```
we can extract the repeated code that decodes the content and returns the topoCell  in the two if/else conditions into a helper.

rohit-nayak-ps reviewed (approved):
❤️

mattlord commented:
Thank you for the review, @rohit-nayak-ps ! I've addressed your comments here: https://github.com/vitessio/vitess/pull/15933/commits/3ff727c2b95cc64f38634c438277654e791a18d7

Thank you for the nudge! I wasn't happy with that function either and it gave me the push to refactor it. 🙂

deepthi commented on go/vt/vtctl/grpcvtctldserver/server.go:
```diff
@@ -2259,24 +2260,26 @@ func (s *VtctldServer) GetTopologyPath(ctx context.Context, req *vtctldatapb.Get
 	defer span.Finish()
 
-	// handle toplevel display: global, then one line per cell.
-	if req.Path == "/" {
```
nit: proto fields are all public by design, and it is fine to access them directly. There is no need to change all accesses to use a getter instead.

mattlord replied:
I've gotten into the habit of using the proto getters as they are nil safe. In this case, for example, if req is nil it will return "" rather than have a nil pointer error.
